Userland/Libraries/LibWeb/DOM/Element.cpp

```diff
@@ -1001,15 +1001,73 @@ void Element::set_scroll_top(double y)
     block_container->set_scroll_offset(scroll_offset);
 }
 
+// https://w3c.github.io/csswg-drafts/cssom-view/#dom-element-scrollwidth
 int Element::scroll_width() const
 {
-    dbgln("FIXME: Implement Element::scroll_width() (called on element: {})", debug_description());
+    // 1. Let document be the element's node document.
+    auto const& document = this->document();
+
+    // 2. If document is not the active document, return zero and terminate these steps.
+    if (!document.is_active())
+        return 0;
+
+    // 3. Let viewport width be the width of the viewport excluding the width of the scroll bar,
+    //    if any, or zero if there is no viewport.
+    auto const* browsing_context = document.browsing_context();
+    auto const viewport_width = browsing_context == nullptr ? 0 : browsing_context->viewport_rect().width();
+
+    // 4. If the element is the root element and document is not in quirks mode
+    //    return max(viewport scrolling area width, viewport width).
+    if (is<HTML::HTMLHtmlElement>(*this) && !document.in_quirks_mode())
+        return max(browsing_context->viewport_scroll_offset().x(), viewport_width);
+
+    // 5. If the element is the body element, document is in quirks mode
+    //    and the element is not potentially scrollable,
+    //    return max(viewport scrolling area width, viewport width).
+    if (is<HTML::HTMLBodyElement>(*this) && document.in_quirks_mode() && !this->is_potentially_scrollable())
+        return max(browsing_context->viewport_scroll_offset().x(), viewport_width);
```

Comment:
Hmm, is `viewport scrolling area width` in the spec really the same as the viewport scroll offset X position? That sounds strange.

Reply:
Oh yeah I think I made a mistake here. I'm confused — does viewport scrolling area calculation also depend on CSS overflow? I see that we do calculate the overflow for a block formatting context (which is what the root element should be), but I'm not sure I understand where we're saving that information.